File: COMP9041/lab07/frequency.py
#!/usr/bin/env python

import sys
import re
import glob

def get_counts(file, pattern_word):
    pattern_word = pattern_word.lower()
    total_count = 0
    word_count = 0

    with open(file, 'r') as f:
        for line in f.readlines():
            pattern_head = re.compile("^[^A-Za-z]+")
            line = re.sub(pattern_head, '', line)
            pattern_split = re.compile("[^A-Za-z]+")
            splitted_line = re.split(pattern_split, line)
            total_count += len(splitted_line) - 1
            for word in splitted_line[:-1]:
                word = word.lower()
                if re.match(re.compile("^" + pattern_word + "$"), word):
                    word_count += 1

    return total_count, word_count

pattern_word = sys.argv[1]

for file in sorted(glob.glob("lyrics/*.txt")):
    total_count, word_count = get_counts(file, pattern_word)
    freq = word_count * 1.0 / total_count
    pat1 = re.compile("^.*?\/")
    file = re.sub(pat1, '', file)
    pat2 = re.compile("\.txt")
    file = re.sub(pat2, '', file)
    pat3 = re.compile("_")
    file = re.sub(pat3, ' ', file)
    print "{:4d}/{:6d} = {:.9f} {}".format(word_count, total_count, freq, file)

File: COMP9041/lab07/frequency.pl
#!/usr/bin/perl -w

# prints the frequency with each artist uses a word specified as argument
# use a hash table of hash tables indexed by artist and word to store the word counts.

$pattern_word = $ARGV[0];
foreach $file (glob "lyrics/*.txt") {
    @result = get_counts($file, $pattern_word);
    # print "@result\n";
    $total_count = $result[0];
    $word_count = $result[1];
    $freq = $word_count / $total_count;
    $file =~ s/^.*?\///g;
    $file =~ s/\.txt//g;
    $file =~ s/_/ /g;
    printf "%4d/%6d = %.9f %s\n", $word_count, $total_count, $freq, $file;
}


# input: file, word_pattern
# output: result = (total_count, word_count)
sub get_counts{
    my $file = $_[0];
    my $pattern_word = $_[1];
    $pattern_word = lc($pattern_word);
    my $total_count = 0;
    my $word_count = 0;
    
    open(my $DATA, '<', $file) or die("Can not open file, $!\n");
    while (my $line = <$DATA>){
        $line =~ s/^[^A-Za-z]+//;
        my @splitted_line = split('[^A-Za-z]+', $line);
        my $length = @splitted_line;
        $total_count = $total_count + $length;
        foreach my $word (@splitted_line){
            $word = lc($word);
            if ($word =~ m/^\Q$pattern_word\E$/){ # \Q \E escape
                $word_count++;
            }
        }
    }
    my @result = ();
    push(@result, $total_count);
    push(@result, $word_count);
    return @result;
}
